src/Interpreters/HashJoin.cpp

```diff
@@ -1365,19 +1385,29 @@ NO_INLINE IColumn::Filter joinRightColumns(
     constexpr JoinFeatures<KIND, STRICTNESS> join_features;
 
     size_t rows = added_columns.rows_to_add;
-    IColumn::Filter filter;
     if constexpr (need_filter)
-        filter = IColumn::Filter(rows, 0);
+        added_columns.filter = IColumn::Filter(rows, 0);
 
     Arena pool;
 
     if constexpr (join_features.need_replication)
         added_columns.offsets_to_replicate = std::make_unique<IColumn::Offsets>(rows);
 
     IColumn::Offset current_offset = 0;
-
-    for (size_t i = 0; i < rows; ++i)
+    size_t max_joined_block_rows = added_columns.max_joined_block_rows;
+    size_t i = 0;
+    for (; i < rows; ++i)
     {
+        if constexpr (join_features.need_replication)
+        {
+            if (unlikely(current_offset > max_joined_block_rows))
+            {
+                added_columns.offsets_to_replicate->resize_assume_reserved(i);
```

Comment:
How do we ensure that `offsets_to_replicate` is not `nullptr` and the size is reserved?

Reply:
Yes, we initialize it if `join_features.need_replication = true` https://github.com/ClickHouse/ClickHouse/blob/d764d8d72b477470b635106980c0f030caaabba7/src/Interpreters/HashJoin.cpp#L1394-L1395

And use it only when flag is enabled